core/src/main/scala/cats/data/Chain.scala

```diff
@@ -1029,6 +1046,7 @@ sealed abstract private[data] class ChainInstances extends ChainInstances1 {
 
       def empty[A]: Chain[A] = Chain.nil
       def combineK[A](c: Chain[A], c2: Chain[A]): Chain[A] = Chain.concat(c, c2)
+      override def fromIterableOnce[A](xs: IterableOnce[A]): Chain[A] = Chain.fromSeq(xs.iterator.toSeq)
```

Comment:
nit...

god... I will never stop finding micro issues:

```scala
override def fromIterableOnce[A](xs: IterableOnce[A]): Chain[A]  =
  xs match {
    case s: Seq[A @unchecked] =>
      // Seq is a subclass of IterableOnce, so the type has to be compatible
     Chain.fromSeq(s) // pay O(1) not O(N) cost
   case notSeq =>
     Chain.fromSeq(notSeq.iterator.toSeq)
  }
```

Reply:
Could it make sense to add a `Chain.fromIterableOnce` that does this instead? I don't really know what to make of these `IterableOnce` possibly side-effecty things.

Reply:
sure, we can put it in Chain.

I don't view this as a side effect so much as we are lacking linear or affine types. What we want is to say this method takes ownership of `xs`. If we had that, then iterating it would be safe (assuming it did no side effect, which we make that assumption all the time in every function and method call).

Reply:
note: It would be really cool if scala added ownership types but I think it is a bit complex to get right and ergonomic.